In [1]:
# Hier nichts ändern!
import sys
sys.path.append("../code")
from setup_AB1 import *

<img src = "../figs/cammp.png" style="height:5em; width:auto" align="right"> 

# Netflix | Arbeitsblatt 1: Datenanalyse <i class="fa fa-database" style="font-size:26px"></i>
Auf diesem Arbeitsblatt kannst du dir einen Eindruck von dem Datensatz der Netflix Challenge verschaffen. 

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-pencil-square-o" style="font-size:30px"></i> <br><br> Daten analysieren  </h3>
    In <strong>Aufgabe 1</strong> werden dir verschiedene Tabellen oder Grafiken angezeigt. <br>
    Schaue dir die Tabellen und Grafiken <strong>erst alle in Ruhe an</strong>. Notiere dann die Antworten zu folgenden Fragen in den folgenden Feldern.
    <ol>
        <li>Was ist der beliebteste Film in dem Datensatz?</li>
        <li>Aus welchem Jahr stammen die ältesten Filme in dem Datensatz? Aus welchem Jahr die jüngsten?</li>
        <li>Wie hat sich die Verteilung der Bewertungen im Laufe der Erscheinungsjahre geändert?</li>
        <li>Welcher Film wurde am häufigsten bewertet?</li>
        <li>Bei welchem Film weichen die Bewertungen der einzelnen Nutzer am wenigsten vom durchschnittlichen Rating ab?</li>
    </ol>
</div>

In [2]:
prompt_1_0_1()
prompt_1_0_2()
prompt_1_0_3()
prompt_1_0_4()
prompt_1_0_5()

Output()

Output()

Output()

Output()

Output()

## 1) Darstellung des Netflix-Datensatzes



<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-laptop" style="font-size:30px"></i> <br><br>Grafische Darstellung: Erscheinungsjahr und durchschnittliches Rating </h3>
    Dir wird eine Grafik angezeigt, in der das Erscheinungsjahr auf der x-Achse und das durchschnittliche Rating auf der y-Achse aufgetragen sind. Du kannst entscheiden, ab welchem Erscheinungsjahr die Filme angezeigt werden, indem du den Slider bewegst.
    <hr>
    <i class="fas fa-exclamation-triangle" style="font-size: 20px"></i> <strong>Hinweise zur Grafik:</strong>
    <ul>
        <li>Wenn du mit der Maus über die Grafik fährst, werden dir weitere Infos zu den einzelnen Filmen angezeigt.</li>
        <li>Die Standardabweichung gibt an, wie stark die Werte eines Merkmals (hier Bewertungen) rund um dessen Mittelwert  streuen. Vereinfacht gesagt, gibt die Standardabweichung die durchschnittliche Entfernung aller Bewertungen eines Filmes vom durchschnittlichen Rating  des Filmes an.</li>
    </ul>
</div>

In [3]:
# Hier nichts ändern!
interactive_plot(create_rls_av_std_plot)

interactive(children=(FloatSlider(value=1900.0, description='start_year', max=2005.0, min=1850.0, step=1.0), O…

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-laptop" style="font-size:30px"></i> <br><br>Grafische Darstellung: Erscheinungsjahr und Anzahl Ratings  </h3>
    Dir wird eine Grafik angezeigt, in der das Erscheinungsjahr auf der x-Achse und die Anzahl der Ratings eines jeden Filmes auf der y-Achse aufgetragen sind.
</div>

In [4]:
# Hier nichts ändern!
interactive_plot(create_rls_count_av_plot)

interactive(children=(FloatSlider(value=1900.0, description='start_year', max=2005.0, min=1850.0, step=1.0), O…

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-laptop" style="font-size:30px"></i> <br><br>Tabellarische Übersicht  </h3>
    Führe den folgenden Code aus.<br>
    Dir wird eine Tabelle mit  Infos zu den Filmen des Netflixdatensatzes angezeigt. Du kannst die Tabelle  nach einer beliebigen Spalte sortieren. Lege fest, nach welcher Spalte (z. B. Anzahl Ratings, Erscheinungsjahr ...) sortiert werden soll. 
Trage dazu den Namen der  Spalte für die Variable <strong>sort1</strong> ein. Du kannst außerdem festlegen, ob die Tabelle aufsteigend (<strong>aufsteigend = True</strong>) oder absteigend (<strong>aufsteigend = False</strong>) sortiert werden soll.
    <hr>
    <i class="fas fa-exclamation-triangle" style="font-size: 20px"></i> <strong>Hinweis:</strong> Die Gänsefüßchen <strong>" "</strong> müssen um die Bezeichnung der Spalte stehen bleiben.
</div>

In [5]:
# Sortierung nach Spalte "sort1"
sort1 = "Filmtitel"

# True = aufsteigende Sortierung, False = absteigende Sortierung
aufsteigend = True

# Ab hier nichts ändern!
df_ger.sort_values(by=[sort1], ascending=aufsteigend)

Filmtitel  Anzahl Ratings  durchschnittl. Rating  \
4090               ... And God Spoke             180                 2.8722   
7755          ...And Justice for All            6350                 3.6717   
8574         ...And Then Came Summer            1397                 2.7316   
9332                 .Com for Murder             153                 2.3203   
5763   .hack//Legend of the Twilight             736                 3.9226   
...                              ...             ...                    ...   
17630                           Zulu            4123                 3.8394   
10084                       Zus & Zo            1184                 2.9274   
15254                       e-Dreams            1812                 2.9752   
2786                        s-Cry-ed             817                 4.1053   
15423       sex, lies, and videotape           28129                 3.4123   

       Standardabweichung  Erscheinungsjahr  
4090               1.2738              1993  
7755                0.892              1979  
8574               1.2044              2000  
9332               1.0886              2002  
5763                1.116              2003  
...                   ...               ...  
17630              0.9635              1964  
10084              0.9871              2002  
15254              1.0745              2002  
2786               0.9827              2003  
15423              0.9694              1989  

[17770 rows x 5 columns]

<i class="fas fa-exclamation-triangle" style="font-size: 20px"></i> <strong>Hinweis zur Tabelle:</strong> Jeder Film hat eine eindeutige Kennzeichnung bzw. Nummer: seine ID.


# 2) Die Rating-Tabelle
Wir werden nun alle Bewertungen, die die Nutzer für die Filme abgegeben haben, in einer Art Tabelle bzw. in einem rechteckigen Zahlenschema notieren. 
Jede Zeile steht für einen User und jede Spalte für einen Film. Die Einträge der Tabelle geben an, welche Bewertung ein Nutzer für einen Film abgegeben hat. Jedoch sind nicht alle Bewertungen bekannt.

Ein Auszug aus der Tabelle könnte  wie folgt aussehen:

<div style="text-align:center">
<img src="../figs/ratingmatrix_example.png" width="500"/> 
 </div>
 
Der Netflixdatensatz besteht aus:
* 17770 Filmen 
<i class="fa fa-film"></i>  
* 480189 Nutzern <i class="fa fa-user"></i>  
* 100480507 Bewertungen (engl. *ratings*) von 1 - 5 <i class="fa fa-star"></i> <i class="fa fa-star"></i> <i class="fa fa-star"></i> <i class="fa fa-star"></i> <i class="fa fa-star"></i>

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-pencil-square-o" style="font-size:30px"></i> <br><br> a) Brainstorming  </h3>
    Sammle Ideen, wie man die <strong>bekannten Bewertungen aus der Bewertungstabelle</strong> (auch <strong>Rating-Tabelle</strong> genannt) nutzen kann, um einem User Filme zu empfehlen, die er noch nicht gesehen hat - und zwar so, dass die Filme dem User vermutlich auch gefallen.
</div>

In [15]:
promt_1_2_a()

Output()

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-pencil-square-o" style="font-size:30px"></i> <br><br> b) Zusammenhänge in den Bewertungsdaten  </h3>
    Schaue dir die folgende beispielhafte <strong>Rating-Tabelle</strong> an. Wir haben dabei ein Beispiel gewählt, bei dem wir den Großteil der Bewertungen bereits kennen.
    <div style="text-align:center">
        <img src="../figs/AB1_A1_a.png" width="500", class="center">
    </div>
    <a id='A1_Teila'></a>
    Welche Zusammenhänge zwischen verschiedenen Usern oder Filmen kannst du in der Rating-Tabelle finden? Beschreibe, was dies für die jeweiligen Nutzer bzw. Film bedeutet.
    <hr>
    <strong>Hinweis:</strong> Es sind mindestens 3 interessante Zusammenhänge zu erkennen
    <hr>
    <a href="../help/help1_1.ipynb"><i class="fas fa-life-ring" style="font-size:20px"></i> &nbsp;Tipp</a>
</div>

In [16]:
promt_1_2_b()

Output()

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-comments" style="font-size:30px"></i> & <i class="fa fa-pencil-square-o" style="font-size:30px"></i> <br><br> c) Vorhersage einzelner Bewertungen  </h3>
    In dem  Beispiel aus <strong>Teil b</strong> hat
    <ul>
        <li>Nutzer 2 den Film 6 nicht gesehen und </li>
        <li>Nutzer 4 den Film 1  nicht gesehen.</li>
    </ul>
    <hr>
    Diskutiere mit deinem Gruppenmitglied: Wie würdet ihr die beiden fehlenden Bewertungen vorhersagen? Und welchen Film sollten wir dem Nutzer dann vorschlagen?
    <h3>Fazit</h3>
    Nachdem wir uns mit dem  Netflixdatensatz vertraut gemacht haben, werden wir auf den nächsten Arbeitsblättern an kleinen Datensätzen ein Empfehlungssystem entwickeln.
</div>

In [17]:
promt_1_2_c()

Output()

<div id="exercise" class="alert alert-info">
    <h3><i class="fa fa-pencil-square-o" style="font-size:30px"></i> <br><br> Zusatzaufgabe für Schnellere:  </h3>
    Wie viel Prozent der Bewertungen (Ratings) des originale Netflixdatensatzes sind bekannt? Wie viel Prozent der Bewertungen sind nicht bekannt? <br>
    <strong>Erinnerung:</strong> Der Netflixdatensatz besteht aus 17770 Filmen, 480189 Nutzern, 100480507 Bewertungen
</div>

In [18]:
promt_1_z()

Output()

# Diskussion im Plenum <img src="../figs/discussion.png" width="50"/>
**Wir werden eure bisherigen Ergebnisse gemeinsam diskutieren!**

Öffne nun das nächste [Arbeitsblatt](./AB2-SuS.ipynb).
 
 ___
<a rel=“license” href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Lizenzvertrag" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a> Dieses Werk ist lizenziert unter einer <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Namensnennung - Weitergabe unter gleichen Bedingungen 4.0 International Lizenz</a>

Autorin: Sarah Schönbrodt